Add support for dns-rfc2136 plugin

Rathios opened the pull request:
<!--
Thank you for contributing to this project!

- This project has a Contributor Code of Conduct: https://voxpupuli.org/coc/
- Please check that here is no existing issue or PR that addresses your problem.
- Our vulnerabilities reporting process is at https://voxpupuli.org/security/

-->
#### Pull Request (PR) description
Adds support for issuing certificates with the Let's Encrypt dns-rfc2136 plugin. Tested on CentOS7.

Three minor bonuses:
- Fixes a small spec bug where `dns-cloudflare` was not being allowed as a value for the Letsencrypt::Plugin type.
- Cleans up some old Puppet 3 syntax
- README gets a header upgrade for better readability (and so usage of this plugin can be documented sanely)

The module already has a dependency on puppetlabs/inifile, but I've used an ini template for (imo) cleaner code. Something to evaluate as well.

#### This Pull Request (PR) fixes the following issues
n/a

bastelfreak commented on manifests/certonly.pp:
```diff
@@ -67,18 +67,35 @@
-    $command_domains  = "--cert-name ${title} -d ${_command_domains}"
+
+    'dns-rfc2136': {
+      require ::letsencrypt::plugin::dns_rfc2136
```
please don't add leading `::` in front of class names or variables. That is considered legacy syntax. It was required for Puppet 3 support.

Rathios replied:
Good to know. There are 3 usages of it at the bottom of the file not related to my code. Update those at the same time?

There's also a couple of `${::path}` usages for execs, but those seem to require it to avoid local scope. Curious if that is an acceptable exception?

Rathios replied:
Hrm. Those should probably be `$facts['path']` instead.

bastelfreak replied:
correct. For facts, `$facts['path']` should be used.

bastelfreak commented on manifests/plugin/dns_rfc2136.pp:
```diff
@@ -0,0 +1,59 @@
+  String $server,
+  String $key_name,
+  String $key_secret,
+  String $key_algorithm = $letsencrypt::params::dns_rfc2136_algorithm,
```
please align all the `=` in this section and enforce the minimal string length for the parameters:
`String[1]`

bastelfreak commented on manifests/plugin/dns_rfc2136.pp:
```diff
@@ -0,0 +1,59 @@
+  String $key_name,
+  String $key_secret,
+  String $key_algorithm = $letsencrypt::params::dns_rfc2136_algorithm,
+  String $port          = $letsencrypt::params::dns_rfc2136_port,
```
We should probably use this datatype? https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/types/port.pp

alexjfisher replied:
And `Stdlib::Host` for `$server`?

Rathios replied:
Integer for `$propagation_seconds` probably wouldn't hurt as well.

bastelfreak commented on manifests/plugin/dns_rfc2136.pp:
```diff
@@ -0,0 +1,59 @@
+  String $key_algorithm = $letsencrypt::params::dns_rfc2136_algorithm,
+  String $port          = $letsencrypt::params::dns_rfc2136_port,
+  Boolean $manage_package = $letsencrypt::params::dns_rfc2136_manage_package,
+  String $config_dir      = $letsencrypt::config_dir,
```
I assume this is always an absolute path? https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/types/absolutepath.pp

bastelfreak commented:
Hi @Rathios, thanks for the PR. IMO a template is fine if we manage the whole file. I assume it isn't populated by any package? Or does certbot ship it by default?

Rathios commented:
Should be all good now.

Rathios commented:
... definitely done now. Curse you, tiny polishable details.

bastelfreak commented:
@Rathios can you take a look at the failing spec tests?

Rathios commented:
Rebased master and added new switch `--rsa-key-size 4096` to spec test, introduced by https://github.com/voxpupuli/puppet-letsencrypt/pull/165. Looks OK now. Thanks for the heads up.

Rathios commented:
Thanks!
